haystack/pipeline.py

```diff
@@ -832,3 +862,198 @@ def run(self, **kwargs):
+                else:
+                    raise KeyError(f"Root node '{root_node}' is invalid. Available options are 'Query' and 'File'.")
+
+            name = node_config["name"]
+            component_type = definitions[name]["type"]
+            component_class = BaseComponent.get_subclass(component_type)
+            replicas = next(comp for comp in data["components"] if comp["name"] == name).get("replicas", 1)
+            handle = cls._create_ray_deployment(name=name, pipeline_config=data, replicas=replicas)
+            pipeline._add_ray_deployment_in_graph(
+                handle=handle,
+                name=name,
+                outgoing_edges=component_class.outgoing_edges,
+                inputs=node_config.get("inputs", []),
+            )
+
+        return pipeline
+
+    @classmethod
+    def _create_ray_deployment(cls, name, pipeline_config, replicas=1):
+        RayDeployment = serve.deployment(_RayDeploymentWrapper, name=name, num_replicas=replicas)
+        RayDeployment.deploy(pipeline_config, name)
+        handle = RayDeployment.get_handle()
+        return handle
+
+    def run(self, **kwargs):
+        has_next_node = True
+        current_node_id = self.root_node
+        input_dict = {"root_node": self.root_node, **kwargs}
+        output_dict = None
+
+        while has_next_node:
+            output_dict, stream_id = ray.get(self.graph.nodes[current_node_id]["component"].remote(**input_dict))
+            input_dict = output_dict
+            next_nodes = self.get_next_nodes(current_node_id, stream_id)
+
+            if len(next_nodes) > 1:
+                join_node_id = list(nx.neighbors(self.graph, next_nodes[0]))[0]
+                if set(self.graph.predecessors(join_node_id)) != set(next_nodes):
+                    raise NotImplementedError(
+                        "The current pipeline does not support multiple levels of parallel nodes."
+                    )
+                inputs_for_join_node = {"inputs": []}
+                for n_id in next_nodes:
+                    output = self.graph.nodes[n_id]["component"].run(**input_dict)
+                    inputs_for_join_node["inputs"].append(output)
+                input_dict = inputs_for_join_node
+                current_node_id = join_node_id
+            elif len(next_nodes) == 1:
+                current_node_id = next_nodes[0]
+            else:
+                has_next_node = False
+
+        return output_dict
+
+    def add_node(self, component, name: str, inputs: List[str]):
+        raise NotImplementedError(
+            "The current implementation of RayPipeline only supports loading Pipelines from a YAML file."
+        )
+
+    def _add_ray_deployment_in_graph(self, handle, name: str, outgoing_edges: int, inputs: List[str]):
+        """
+        Add the Ray deployment handle in the Pipeline Graph.
+
+        :param handle: Ray deployment `handle` to add in the Pipeline Graph.
```

Comment:
What type is this `handle`? Can we add an example or link ray docs if it is a standardized concept on their side?